### PR TITLE
Add How-to Refresh Certificates

### PR DESCRIPTION
**Pull request by mateoflorido**
## Overview
This pull request introduces a how-to guide for refreshing certificates in a Canonical Kubernetes cluster.

### CODE REVIEW

File: docs/src/snap/howto/refresh-certs.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+```
+This command refreshes the certificates for the worker node. The `--expires-in`
+flag specifies the certificate's validity period. As mentioned in the control
+plane section, this duration can be set using any unit accepted by the
+[ParseDuration][] function in Go, in addition to years, months, or days.
+
+3. During this process, multiple Certificate Signing Requests (CSRs) are created.
+The command output will guide you on how to approve the CSRs on any control plane
+node in the cluster.
+```bash
```
Comment (HomayoonAlimohammadi) [inline docs/src/snap/howto/refresh-certs.md:63]: using `bash` looks like the right thing to do here, but I noticed some weird coloring in the generated docs. like highlighting the `for` and some numbers in other blocks.

**Review by HomayoonAlimohammadi — COMMENTED** [review]
Thanks a lot @mateoflorido! Just a minor comment, otherwise LGTM.

File: docs/src/snap/howto/refresh-certs.md (hunk 1 above)
Comment (bschimke95) [reply docs/src/snap/howto/refresh-certs.md:63]: According to Nick, you may just remove this and ReadTheDocs will fall-back to something sane.

File: docs/src/snap/howto/refresh-certs.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+services, and potential security risks. This how-to will walk you through
+the steps to refresh the certificates for both control plane and worker
+nodes in your {{product}} cluster.
+
+## What you will need
+- A running {{product}} cluster
+
+```{note} To refresh the certificates in your cluster, make sure it was
+initially set up with self-signed certificates or with your own CA certificates
+and keys during the bootstrap process.
```
Comment (bschimke95) [inline docs/src/snap/howto/refresh-certs.md:16]: I don't think we currently can rotate certificates that are signed by external CA, can we? 
In the proposal, this was out of scope: https://github.com/canonical/k8s-snap/blob/main/docs/proposals/002-refresh-certificates.md#external-ca-certificate

File: docs/src/snap/howto/refresh-certs.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+```
+
+### Refreshing Control Plane Node Certificates
+To refresh the certificates on control plane nodes, follow these steps:
+
+1. Access each control plane node in your cluster.
```
Comment (bschimke95) [inline docs/src/snap/howto/refresh-certs.md:22]: Maybe add the following on top instead:
`Perform the following steps on each control-plane node of the cluster.`

File: docs/src/snap/howto/refresh-certs.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+```
+
+This command refreshes the certificates for the control plane node, adding an
+extra Subject Alternative Name (SAN) to the certificate. The `--expires-in`
+flag sets the certificate's validity duration, which can be specified in years,
+months, days, or any unit accepted by the [ParseDuration][] function in Go.
+
+```{note} Ensure that you provide the same SANs that were used when the cluster
+was bootstrapped. If you don't, the control plane may fail to communicate with
+other nodes in the cluster.
```
Comment (bschimke95) [inline docs/src/snap/howto/refresh-certs.md:36]: I'm curious. Why does the user need to provide the extra-sans if they are the same as the bootstrap sans?

File: docs/src/snap/howto/refresh-certs.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+This command refreshes the certificates for the worker node. The `--expires-in`
+flag specifies the certificate's validity period. As mentioned in the control
+plane section, this duration can be set using any unit accepted by the
+[ParseDuration][] function in Go, in addition to years, months, or days.
+
+3. During this process, multiple Certificate Signing Requests (CSRs) are created.
+The command output will guide you on how to approve the CSRs on any control plane
+node in the cluster.
+```bash
+root@w-1:~# k8s refresh-certs --expires-in 10y --timeout 10s
```
Comment (bschimke95) [inline docs/src/snap/howto/refresh-certs.md:64]: nit: commands and output should always be separated so that users can easily copy the commands. For the same reason, I'd remove the `root@w-1:#` prefix

File: docs/src/snap/howto/refresh-certs.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+This command approves the CSRs, allowing the certificates to be created.
+
+5. After approving all the requested CSRs, the worker node will automatically
+refresh its the certificates and restart the necessary services. You should see
+a confirmation message similar to the following:
+```bash
+Certificates have been successfully refreshed, and will expire at 2034-08-27 21:00:00 +0000 UTC.
+```
+
+## Summing Up
```
Comment (bschimke95) [inline docs/src/snap/howto/refresh-certs.md:90]: I don't think we need this section

**Review by bschimke95 — COMMENTED** [review]
looking great, only a bunch of small comments.

File: docs/src/snap/howto/refresh-certs.md (hunk 2 above)
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:16]: Technically, we can. In the bootstrap configuration, you can provide all the [certificates](https://github.com/canonical/k8s-snap-api/blob/main/api/v1/type_bootstrap_config.go#L27-L32) for the cluster, including the CA cert and CA key, which allow the cluster to manage its own certificates. This is possible if the user provides all the necessary custom certificates. However, if it causes more confusion, I’ll remove that part and leave only the self-signed note.

File: docs/src/snap/howto/refresh-certs.md (hunk 4 above)
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:36]: We are not storing the bootstrap SANs, as these are node-specific. The user might want to change the extra SANs when refreshing the certificates, and for orchestration layers (CAPI or Juju), they already know about the cluster-wide extra SANs we want to add.

File: docs/src/snap/howto/refresh-certs.md (hunk 1 above)
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:63]: Sounds good, let me update the code block.

File: docs/src/snap/howto/refresh-certs.md (hunk 5 above)
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:64]: Sure, sounds good

File: docs/src/snap/howto/refresh-certs.md (hunk 6 above)
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:90]: Cool, let's remove it. I noticed a couple of Summing Up sections in other how-tos, so I wanted to swim with the current 😄

File: docs/src/snap/howto/refresh-certs.md (hunk 2 above)
Comment (bschimke95) [reply docs/src/snap/howto/refresh-certs.md]: True, I was messing that up with CA rotation. It's fine to keep it

File: docs/src/snap/howto/refresh-certs.md (hunk 4 above)
Comment (bschimke95) [reply docs/src/snap/howto/refresh-certs.md:38]: In the comment you write `The user might want to change the extra SANs when refreshing the certificates` but the text says: `Ensure that you provide the same SANs that were used when the cluster was bootstrapped.` Maybe at an "Ensure that you provide at least..."

File: docs/src/snap/howto/refresh-certs.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+This command refreshes the certificates for the worker node. The `--expires-in`
+flag specifies the certificate's validity period. As mentioned in the control
+plane section, this duration can be set using any unit accepted by the
+[ParseDuration][] function in Go, in addition to years, months, or days.
+
+3. During this process, multiple Certificate Signing Requests (CSRs) are
+created. The command output will guide you on how to approve the CSRs on any
+control plane node in the cluster.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:69]:
```suggestion
3. During the certificate refresh, multiple Certificate Signing Requests (CSRs) are
created. Follow the instructions in the command output to approve the CSRs on any
control plane node in the cluster.
```

File: docs/src/snap/howto/refresh-certs.md (hunk 6 above)
Comment (bschimke95) [reply docs/src/snap/howto/refresh-certs.md]: Up to you. IMHO this sounded too much like a summary that an AI would generate at the end of a prompt response, hence I wouldn't add it here. (In the end, our docs are hand-crafted with sweat and tears :D )

File: docs/src/snap/howto/refresh-certs.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+```
+sudo k8s refresh-certs --expires-in 10y --timeout 10s
+The following CertificateSigningRequests should be approved. Run the following commands on any of the control plane nodes of the cluster:
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kubelet-serving
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kubelet-client
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kube-proxy-client
+Waiting for certificates to be created...
+```
+
+4. On any control plane node, run the commands provided in the output:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:80]:
```suggestion
4. Approve the CSRs by running the commands from the previous output on any control plane node:
```

File: docs/src/snap/howto/refresh-certs.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+certificatesigningrequest.certificates.k8s.io/k8sd-3974895791729870959-worker-kubelet-serving approved
+certificatesigningrequest.certificates.k8s.io/k8sd-3974895791729870959-worker-kubelet-client approved
+certificatesigningrequest.certificates.k8s.io/k8sd-3974895791729870959-worker-kube-proxy-client approved
+```
+
+This command approves the CSRs, allowing the certificates to be created.
+
+5. After approving all the requested CSRs, the worker node will automatically
+refresh its the certificates and restart the necessary services. You should see
+a confirmation message similar to the following:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:95]:
```suggestion
5. After approving all the requested CSRs, the worker node will automatically
refresh its certificates and restart the necessary services:
```

**Review by louiseschmidtgen — COMMENTED** [review]
Thank you @mateoflorido! 
Great work on the cert refreshes, added a few suggestions to improve the flow.

File: docs/src/snap/howto/refresh-certs.md (hunk 4 above)
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:38]: @louiseschmidtgen It's possible to check the current SANs. Let's add a snippet to check this!
@bschimke95 Your are absolutely right, let me rephrase that to convey that the user should include the host specific SANs.

File: docs/src/snap/howto/refresh-certs.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+### Refreshing Worker Node Certificates
+
+To refresh the certificates on worker nodes, perform the following steps on
+each worker node in your cluster:
+
+1. Access each worker node in your cluster.
+
+2. Run the `refresh-certs` command:
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:66]: The text already indicates that the steps should be run on each worker node. Would you like me to improve the wording or restructure the paragraph?

File: docs/src/snap/howto/refresh-certs.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+services, and potential security risks. This how-to will walk you through
+the steps to refresh the certificates for both control plane and worker
+nodes in your {{product}} cluster.
+
+## What you will need
+
+- A running {{product}} cluster
+
+```{note} To refresh the certificates in your cluster, make sure it was
+initially set up with self-signed certificates during the bootstrap process.
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply docs/src/snap/howto/refresh-certs.md:16]: I think this will only add complexity to the guide. Without using `openssl` tricks, it would be difficult to determine if the certificates are self signed or not. The administrator should already know if the cluster was setup with custom CA, as they should have provided custom certs during the bootstrap and join phases.

[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply docs/src/snap/howto/refresh-certs.md:16]: That makes sense, thanks Mateo!

File: docs/src/snap/howto/refresh-certs.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+## Prerequisites
+
+- A running {{product}} cluster
+
+```{note} To refresh the certificates in your cluster, make sure it was
+initially set up with self-signed certificates during the bootstrap process.
+```
+
+### Refreshing Control Plane Node Certificates
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:19]:
```suggestion
### Refresh Control Plane Node Certificates
```

File: docs/src/snap/howto/refresh-certs.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+To refresh the certificates on control plane nodes, perform the following steps
+on each control plane node in your cluster:
+
+1. Access each control plane node in your cluster.
+2. Run the `refresh-certs` command:
+
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:26]:
```suggestion
```
We can remove these steps now that we've said this in the sentence above.

File: docs/src/snap/howto/refresh-certs.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+```
+sudo k8s refresh-certs --expires-in 1y --extra-sans mynode.local
+```
+
+This command refreshes the certificates for the control plane node, adding an
+extra [Subject Alternative Name][] (SAN) to the certificate. Checking the
+current SANs on your node can be done by running the following command:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:33]:
```suggestion
extra [Subject Alternative Name][] (SAN) to the certificate. Check the
current SANs on your node by running the following command:
```

File: docs/src/snap/howto/refresh-certs.md [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+```
+openssl x509 -in /etc/kubernetes/pki/apiserver.crt -noout -text | grep -A 1 "Subject Alternative Name"
+```
+
+```{note} If your node setup includes additional SANs, be sure to provide the
+specific SANs for each node as needed using the `--extra-sans` flag. While this
+isn't required, omitting them could impact your node's ability to communicate
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:41]:
```suggestion
is not required, omitting them could impact your node's ability to communicate
```

File: docs/src/snap/howto/refresh-certs.md [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+with other components in the cluster.
+```
+
+The `--expires-in`flag sets the certificate's validity duration, which can
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:45]:
```suggestion
The `--expires-in` flag sets the certificate's validity duration, which can
```

File: docs/src/snap/howto/refresh-certs.md [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+3. The cluster will automatically update the certificates in the control plane
+node and restart the necessary services. The new expiration date will be
+displayed in the command output:
+
+```
+Certificates have been successfully refreshed, and will expire at 2025-08-27 21:00:00 +0000 UTC.
+```
+
+### Refreshing Worker Node Certificates
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:57]:
```suggestion
### Refresh Worker Node Certificates
```

File: docs/src/snap/howto/refresh-certs.md [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+To refresh the certificates on worker nodes, perform the following steps on
+each worker node in your cluster:
+
+1. Access each worker node in your cluster.
+
+2. Run the `refresh-certs` command:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:64]:
```suggestion
```

File: docs/src/snap/howto/refresh-certs.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply docs/src/snap/howto/refresh-certs.md:64]: Your sentences are good but we can consolidate them into the first one to avoid duplicating information.

File: docs/src/snap/howto/refresh-certs.md [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+sudo k8s refresh-certs --expires-in 10y --timeout 10s
+The following CertificateSigningRequests should be approved. Run the following commands on any of the control plane nodes of the cluster:
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kubelet-serving
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kubelet-client
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kube-proxy-client
+Waiting for certificates to be created...
+```
+
+4. Approve the CSRs by running the commands from the previous output on any
+control plane node:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:89]:
```suggestion
4. Approve the CSRs by running the following commands on any control plane node, which allows the certificates to be created:
```

File: docs/src/snap/howto/refresh-certs.md [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+```
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kubelet-serving
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kubelet-client
+k8s kubectl certificate approve k8sd-3974895791729870959-worker-kube-proxy-client
+certificatesigningrequest.certificates.k8s.io/k8sd-3974895791729870959-worker-kubelet-serving approved
+certificatesigningrequest.certificates.k8s.io/k8sd-3974895791729870959-worker-kubelet-client approved
+certificatesigningrequest.certificates.k8s.io/k8sd-3974895791729870959-worker-kube-proxy-client approved
+```
+
+This command approves the CSRs, allowing the certificates to be created.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:100]:
```suggestion
```
Moving this bit into the sentence above the command.

File: docs/src/snap/howto/refresh-certs.md [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+```{note} If your node setup includes additional SANs, be sure to provide the
+specific SANs for each node as needed using the `--extra-sans` flag. While this
+isn't required, omitting them could impact your node's ability to communicate
+with other components in the cluster.
+```
+
+The `--expires-in`flag sets the certificate's validity duration, which can
+be specified in years, months, days, or any unit accepted by the
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/refresh-certs.md:46]:
```suggestion
be specified in years, months, days, or any other unit accepted by the
```

**Review by louiseschmidtgen — COMMENTED** [review]
One last round of comments then we are good to go!

**Review by bschimke95 — COMMENTED** [review]
LGTM content-wise.
Leaving the wording LGTM to people who know how to put words into an order that makes sense and sounds nice, like @louiseschmidtgen and @evilnick  :)

**Review by bschimke95 — APPROVED** [review]

**Review by louiseschmidtgen — APPROVED** [review]
Wonderful, thanks @mateoflorido!
